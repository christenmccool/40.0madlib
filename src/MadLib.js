import React, {useState} from "react";
import "./MadLib.css";
import Button from 'react-bootstrap/Button';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';

import MadLibForm from "./MadLibForm";
import MadLibSelect from "./MadLibSelect";


/**  
 * MadLib Component renders all child components.
 * Intial rendering is of "MadLib" title and MadLibSelect form to choose template.
 * Once template is selected, MadLibForm is rendered for user to input story words.
 * Words from form construct the story which is displayed to user.
 * Options to reset with "Restart" and "Cancel" buttons
*/
const MadLib = ({choices=defaultChoices}) => {
  const [choice, setChoice] = useState(null);
  const [story, setStory] = useState(null);

  // Choose title/template to be used for story construction from array of choices
  const makeChoice = (idx) => {
    setChoice(choices[idx]);
  }

  // Write story given template in state given words object
  const writeStory = (words) => {
    setStory(words);

    const storyWords = choice.template.map(ele => {
      if (choice.prompts.includes(ele)) {
        return words[ele];
      } else {
        return ele;
      }
    });
    setStory(storyWords.join(" ")+".");
  }

  // Reset choice and story state variables, causing a re-rendering to the inital screen
  const reset = () => {
    setChoice(null);
    setStory(null);
  }

  const renderMain = () => {
    // If a story is completed, display the story and restart buttons
    if (story) {
      return (
        <>
          <h2>{choice.title}</h2>
          <p> {story} </p>
          <Button onClick={reset}>Restart</Button>
        </>
      )
    // If a template has been chosen but the story is not yet complete, display the input form and cancel button
    } else if (choice) {
      return (
        <>
          <Row>
            <MadLibForm prompts={choice.prompts} writeStory={writeStory} />
          </Row>
          <Row>
            <Col>
              <Button className="btn-secondary" onClick={reset}>Cancel</Button>
            </Col>
          </Row>
        </>
      )
    // If no template has been chosen, display the choice select form
    } else {
      return <MadLibSelect titles={choices.map(ele => ele.title)} makeChoice={makeChoice}/>;
    }
  }
  return (
    <Container className="MadLib">
      <Row className="m-3 justify-content-center">
        <Col>
          <h1 className="MadLib-heading">Madlibs!</h1>
          {renderMain()}
        </Col>
      </Row>
    </Container>
  )
}

// Template should include each elememnt of prompts
// Prompts must be unique
const defaultChoices = [
  {title: "A love story",
   template: ["There was a", "adjective", "noun1", "who loved a", "color", "noun2"],
   prompts: ["noun1", "noun2", "adjective", "color"]
  },
  {title: "Animal escapades",
   template: ["My", "adjective1", "friend named", "name", "went on a trip to", "place", "with a", "adjective2", "animal"],
   prompts: ["name", "animal", "place", "adjective1", "adjective2"]
  },
  {title: "Short but sweet",
  template: ["The", "adjective", "singular noun", "plural verb"],
  prompts: ["singular noun", "adjective", "plural verb"]
 }
]

export default MadLib;