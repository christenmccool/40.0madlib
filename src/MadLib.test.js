import { render, fireEvent } from '@testing-library/react';
import MadLib from './MadLib';

test('renders without crashing', () => {
  render(<MadLib />);
});

test("if matches snapshot", () => {
  const {asFragment} = render(<MadLib />);
  expect(asFragment()).toMatchSnapshot();
})

test("if select form is on initial rendering of page", () => {
  const {queryByText} = render(<MadLib />);

  expect(queryByText("Choose a MadLib:")).toBeInTheDocument();
  expect(queryByText("Get Story")).not.toBeInTheDocument();
})

test("if select form causes a rending of input form", () => {
  const {queryByText, debug} = render(<MadLib />);

  const selectForm = queryByText("Choose a MadLib:").closest("select");
  const selectBtn = queryByText("Select");

  fireEvent.change(selectForm, {target: {value: 'option0'}});
  fireEvent.click(selectBtn);

  expect(queryByText("Choose a MadLib:")).not.toBeInTheDocument();
  expect(queryByText("Get Story")).toBeInTheDocument();
})

test("if input form causes a rending of story", () => {
  const {queryByText, queryByPlaceholderText, debug} = render(<MadLib />);

  const selectForm = queryByText("Choose a MadLib:").closest("select");
  const selectBtn = queryByText("Select");
  fireEvent.change(selectForm, {target: {value: 'option0'}});
  fireEvent.click(selectBtn);

  const noun1 = queryByPlaceholderText("noun1");
  const noun2 = queryByPlaceholderText("noun2");
  const adjective = queryByPlaceholderText("adjective");
  const color = queryByPlaceholderText("color");
  const submitBtn = queryByText("Get Story");


  fireEvent.change(noun1, { target: { value: 'dog' } });
  fireEvent.change(noun2, { target: { value: 'cat' } });
  fireEvent.change(adjective, { target: { value: 'loud' } });
  fireEvent.change(color, { target: { value: 'blue' } });
  fireEvent.click(submitBtn);

  expect(queryByText("Choose a MadLib:")).not.toBeInTheDocument();
  expect(queryByText("Get Story")).not.toBeInTheDocument();
  expect(queryByText("There was a loud dog who loved a blue cat.")).toBeInTheDocument();
  expect(queryByText("Restart")).toBeInTheDocument();
})

test("if cancel button resets page", () => {
  const {queryByText, queryByPlaceholderText, debug} = render(<MadLib />);

  const selectForm = queryByText("Choose a MadLib:").closest("select");
  const selectBtn = queryByText("Select");
  fireEvent.change(selectForm, {target: {value: 'option0'}});
  fireEvent.click(selectBtn);

  const noun1 = queryByPlaceholderText("noun1");
  const noun2 = queryByPlaceholderText("noun2");
  const adjective = queryByPlaceholderText("adjective");
  const color = queryByPlaceholderText("color");
  const submitBtn = queryByText("Get Story");
  fireEvent.change(noun1, { target: { value: 'dog' } });
  fireEvent.change(noun2, { target: { value: 'cat' } });
  fireEvent.change(adjective, { target: { value: 'loud' } });
  fireEvent.change(color, { target: { value: 'blue' } });

  const cancelBtn = queryByText("Cancel");
  fireEvent.click(cancelBtn);

  expect(queryByText("Choose a MadLib:")).toBeInTheDocument();
  expect(queryByText("Get Story")).not.toBeInTheDocument();
  expect(queryByText("There was a loud dog who loved a blue cat.")).not.toBeInTheDocument();
  expect(queryByText("Restart")).not.toBeInTheDocument();
})

test("if restart button resets page", () => {
  const {queryByText, queryByPlaceholderText, debug} = render(<MadLib />);

  const selectForm = queryByText("Choose a MadLib:").closest("select");
  const selectBtn = queryByText("Select");
  fireEvent.change(selectForm, {target: {value: 'option0'}});
  fireEvent.click(selectBtn);

  const noun1 = queryByPlaceholderText("noun1");
  const noun2 = queryByPlaceholderText("noun2");
  const adjective = queryByPlaceholderText("adjective");
  const color = queryByPlaceholderText("color");
  const submitBtn = queryByText("Get Story");
  fireEvent.change(noun1, { target: { value: 'dog' } });
  fireEvent.change(noun2, { target: { value: 'cat' } });
  fireEvent.change(adjective, { target: { value: 'loud' } });
  fireEvent.change(color, { target: { value: 'blue' } });
  fireEvent.click(submitBtn);

  const restartBtn = queryByText("Restart");
  fireEvent.click(restartBtn);

  expect(queryByText("Choose a MadLib:")).toBeInTheDocument();
  expect(queryByText("Get Story")).not.toBeInTheDocument();
  expect(queryByText("There was a loud dog who loved a blue cat.")).not.toBeInTheDocument();
  expect(queryByText("Restart")).not.toBeInTheDocument();
})