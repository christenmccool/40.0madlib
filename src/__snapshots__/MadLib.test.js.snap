// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MadLib container"
  >
    <div
      class="m-3 justify-content-center row"
    >
      <div
        class="col"
      >
        <h1
          class="MadLib-heading"
        >
          Madlibs!
        </h1>
        <form
          class="MadLibSelect mt-4"
        >
          <div
            class="justify-content-md-center row"
          >
            <div
              class="col-md-6"
            >
              <select
                class="MadLibSelect form-select"
                id="title"
                name="title"
                required=""
              >
                <option
                  disabled=""
                  value=""
                >
                  Choose a MadLib:
                </option>
                <option
                  value="option0"
                >
                  A love story
                </option>
                <option
                  value="option1"
                >
                  Animal escapades
                </option>
                <option
                  value="option2"
                >
                  Short but sweet
                </option>
              </select>
              <button
                class="m-4 btn btn-primary"
                type="submit"
              >
                Select
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
